src/Orleans/IDs/GrainIdFactory.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using Orleans.Serialization;
+
+namespace Orleans.Runtime
+{
+    internal static class GrainIdFactory
+    {
+        private const int INTERN_CACHE_INITIAL_SIZE = InternerConstants.SIZE_LARGE;
+        private static readonly object lockable = new object();
+        private static readonly TimeSpan internCacheCleanupInterval = InternerConstants.DefaultCacheCleanupFreq;
+        private static Interner<UniqueKey, GrainId> grainIdInternCache;
+
+        public static GrainId NewId()
+        {
+            return FindOrCreateGrainId(UniqueKey.NewKey(Guid.NewGuid(), UniqueKey.Category.Grain));
+        }
+
+        public static GrainId NewClientId(string clusterId = null)
+        {
+            return NewClientId(Guid.NewGuid(), clusterId);
+        }
+
+        internal static GrainId NewClientId(Guid id, string clusterId = null)
```

Comment:
quick question (not necessarily prompting a change): why are these members internal, if the factory type itself is internal?

Reply:
Because I just used resharper refactoring utilities. Yes I guess we could mark these members public in the future

Reply:
yes, that's what I assumed. It is fine though for now anyway, I don't really mind, and the minimal diff is probably better at this point